datafusion/physical-expr/src/expressions/in_list.rs

```diff
@@ -471,37 +450,50 @@ impl InListExpr {
             })
             .collect::<Vec<&str>>();
 
-        if negated {
-            if contains_null {
-                Ok(ColumnarValue::Array(Arc::new(
-                    array
-                        .iter()
-                        .map(|x| match x.map(|v| !values.contains(&v)) {
-                            Some(true) => None,
-                            x => x,
-                        })
-                        .collect::<BooleanArray>(),
-                )))
-            } else {
-                Ok(ColumnarValue::Array(Arc::new(not_in_list_utf8(
-                    array, &values,
-                )?)))
-            }
-        } else if contains_null {
-            Ok(ColumnarValue::Array(Arc::new(
-                array
-                    .iter()
-                    .map(|x| match x.map(|v| values.contains(&v)) {
-                        Some(false) => None,
-                        x => x,
-                    })
-                    .collect::<BooleanArray>(),
-            )))
-        } else {
-            Ok(ColumnarValue::Array(Arc::new(in_list_utf8(
-                array, &values,
-            )?)))
-        }
+        Ok(collection_contains_check!(
+            array,
+            values,
+            negated,
+            contains_null
+        ))
+    }
+
+    fn compare_binary<T: OffsetSizeTrait>(
+        &self,
+        array: ArrayRef,
+        list_values: Vec<ColumnarValue>,
+        negated: bool,
+    ) -> Result<ColumnarValue> {
+        let array = array
+            .as_any()
+            .downcast_ref::<GenericBinaryArray<T>>()
+            .unwrap();
+
+        let contains_null = list_values
+            .iter()
+            .any(|v| matches!(v, ColumnarValue::Scalar(s) if s.is_null()));
+        let values = list_values
+            .iter()
+            .flat_map(|expr| match expr {
+                ColumnarValue::Scalar(s) => match s {
+                    ScalarValue::Binary(Some(v)) | ScalarValue::LargeBinary(Some(v)) => {
+                        Some(v.as_slice())
+                    }
+                    ScalarValue::Binary(None) | ScalarValue::LargeBinary(None) => None,
+                    datatype => unimplemented!("Unexpected type {} for InList", datatype),
+                },
+                ColumnarValue::Array(_) => {
+                    unimplemented!("InList does not yet support nested columns.")
```

Comment:
I think this case is when one of the arguments to `InList` is a column (rather than a constant)

So like `col1 IN ('1', '2', col2)` type thing

(no change needed I was just trying to clarify)